src/Stratis.Bitcoin/P2P/PeerDiscoveryLoop.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.P2P
+{
+    /// <summary> Async loop that discovers new peers to connect to.</summary>
+    public sealed class PeerDiscoveryLoop : IDisposable
+    {
+        /// <summary> The async loop we need to wait upon before we can shut down this connector.</summary>
+        private IAsyncLoop asyncLoop;
+
+        /// <summary> Factory for creating background async loop tasks.</summary>
+        private readonly IAsyncLoopFactory asyncLoopFactory;
+
+        /// <summary>Global application life cycle control - triggers when application shuts down.</summary>
+        private readonly INodeLifetime nodeLifetime;
+
+        private readonly NodeConnectionParameters parameters;
+
+        /// <summary> Peer address manager instance, see <see cref="IPeerAddressManager"/>.</summary>
+        private readonly IPeerAddressManager peerAddressManager;
+
+        /// <summary> The amount of peers to find.</summary>
+        private readonly int peersToFind;
+
+        /// <summary> The network the node is running on.</summary>
+        private readonly Network network;
+
+        public PeerDiscoveryLoop(
+            IAsyncLoopFactory asyncLoopFactory,
+            Network network,
+            NodeConnectionParameters nodeConnectionParameters,
+            INodeLifetime nodeLifetime,
+            IPeerAddressManager peerAddressManager)
+        {
+            Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+
+            this.asyncLoopFactory = asyncLoopFactory;
+            this.parameters = nodeConnectionParameters;
+            this.peerAddressManager = peerAddressManager;
+            this.peersToFind = this.parameters.PeerAddressManagerBehaviour().PeersToDiscover;
+            this.network = network;
+            this.nodeLifetime = nodeLifetime;
+        }
+
+        /// <summary>
+        /// Starts an asynchronous loop that periodicly tries to discover new peers to add to the 
+        /// <see cref="PeerAddressManager"/>.
+        /// </summary>
+        public void Start()
+        {
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.StartAsync), async token =>
```

Comment:
nameof(this.StartAsync) will be StartAsync, that's not very unique name for the loop
as i understand loop names they should be rather unique to be able to recognize which loop is which